src/Serializer.js

```diff
@@ -0,0 +1,68 @@
+/*
+ * Serializer.js - common SPI for serializer plugins
+ *
+ * Copyright © 2024 JEDLSoft
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import NotImplementedError from "./NotImplementedError.js";
+import PipelineElement from "./PipelineElement.js";
+
+/* @typedef {import("IntermediateRepresentation")} IntermediateRepresentation */
+/* @typedef {import("SourceFile")} SourceFile */
+
+/**
+ * @class common SPI for serializer plugins
+ *
+ * A serializer converts an IntermediateRepresentation into a SourceFile instance
+ * that can be written back to disk.
+ *
+ * @abstract
+ */
+class Serializer extends PipelineElement {
+    /**
+     * Construct a new serializer instance.
+     *
+     * @param {Object} [options] options to the constructor
+     * @param {Function} [options.getLogger] a callback function provided by
+     * @param {object} [options.settings] additional settings that can be passed to the serializer
+     * the linter to retrieve the log4js logger
+     */
+    constructor(options) {
+        super(options);
+        if (this.constructor === Serializer) {
+            throw new Error("Cannot instantiate abstract class Plugin directly!");
+        }
+        this.getLogger = options?.getLogger;
+    }
+
+    /**
+     * Serialize the given intermediate representation into a SourceFile instance. The
+     * intermediate representation is an object that represents the parsed
+     * form of the file. The serializer converts this object into a SourceFile
+     * that can be written back to disk.
```

Comment:
This is intentional, yes. The Parser takes a SourceFile and the Serializer creates one. Eventually, the SourceFile can be the abstraction that knows how to read/write the contents but does not know anything else about the contents. It can read from a database, a file on disk, or even a network resource. Maybe it is misnamed now - we could call it something like LinterFile, but I think we're sorta stuck with the name.